skfp/utils/splitters.py

```diff
@@ -0,0 +1,221 @@
+    This method, suggested in the MoleculeNet paper, helps to test the model's ability to
+    generalize to entirely new scaffolds.
+
+    Parameters
+    ----------
+    data : list[str]
+        Sequence representing either SMILES strings or RDKit Mol objects.
+
+    train_size : float, default=0.8
+        The fraction of data to be used for the training split.
+
+    test_size : float, default=0.2
+        The fraction of data to be used for the test split.
+
+    return_indices : bool, default=False
+        Should the method return SMILES/RDKit Mol objects or just indices in the dataset.
+
+    include_chirality: bool, default=False
+        Should the method take chirality of molecules into consideration.
+
+    Returns
+    ----------
+    subsets : tuple[list, list]
+        A tuple of train and test subsets. The format depends on `return_indices`:
+        - If `return_indices` is False, returns lists of SMILES strings or RDKit Mol objects depending
+        on the input.
+        - If `return_indices` is True, returns lists of indices.
+
+    Notes
+    -----
+    - This implementation uses Bemis-Murcko scaffolds (Bemis and Murcko, 1996) to group molecules.
+    - The split is fully deterministic, with the smallest scaffold sets assigned to the test
+    subset and the largest to the training subset.
+    - The split fractions (train_size, test_size) must sum to 1.
+
+    References
+    ----------
+    - Bemis, G. W., & Murcko, M. A. (1996). The properties of known drugs. 1. Molecular frameworks.
+    Journal of Medicinal Chemistry, 39(15), 2887-2893.
+    - Z. Wu, B. Ramsundar, E. N. Feinberg, J. Gomes, C. Geniesse, A. S. Pappu, K. Leswing, V. Pande (2017).
+    MoleculeNet: A Benchmark for Molecular Machine Learning. Chemical Science, 9(2), 513-530.
+    """
+    np.testing.assert_almost_equal(train_size + test_size, 1.0)
+
+    scaffolds = create_scaffolds(list(data), include_chirality)
+    scaffold_sets = sorted(scaffolds.values(), key=len)
+
+    train_cutoff: int = int(train_size * len(data))
+    train_ids: list[int] = []
+    test_ids: list[int] = []
+
+    for scaffold_set in scaffold_sets:
+        if len(train_ids) < train_cutoff:
+            train_ids.extend(scaffold_set)
+        else:
+            test_ids.extend(scaffold_set)
+
+    if return_indices:
+        return train_ids, test_ids
+    else:
+        return _get_data_from_indices(
+            data, scaffolds, train_ids
+        ), _get_data_from_indices(data, scaffolds, test_ids)
+
+
+def scaffold_train_valid_test_split(
+    data: Sequence[Union[str, Mol]],
+    train_size: float = 0.8,
+    test_size: float = 0.1,
+    valid_size: float = 0.1,
+    return_indices: bool = False,
+    include_chirality: bool = False,
+) -> tuple[list, list, list]:
+    """
+    Split a list of SMILES or RDKit Mol objects into train, validation,
+    and test subsets using Bemis-Murcko scaffolds.
```

Comment:
You can make those lines longer, I think. We generally try to get ~88 line length

Reply:
Also, add citations and fixes as in the comment above for train-test function